addArchivedFileSet to ZipArchiver does not work with tar archives

When I try to add the content of tar file to zip file (using `addArchivedFileSet`) I got an exception. I created a test class that reproduces the problem:

```java
package org.codehaus.plexus.archiver;

import java.io.File;

import org.codehaus.plexus.PlexusTestCase;
import org.codehaus.plexus.archiver.tar.TarArchiver;
import org.codehaus.plexus.archiver.tar.TarLongFileMode;
import org.codehaus.plexus.archiver.util.DefaultArchivedFileSet;
import org.codehaus.plexus.archiver.zip.ZipArchiver;
import org.codehaus.plexus.logging.Logger;

public class AddArchivedFileSetTest
    extends PlexusTestCase
{
    public void setUp()
        throws Exception
    {
        super.setUp();
        getContainer().getLoggerManager().setThreshold( Logger.LEVEL_DEBUG );
    }

    public void testTarToZip()
        throws Exception
    {
        TarArchiver tarArchiver = (TarArchiver) lookup( Archiver.ROLE, "tar" );
        tarArchiver.setLongfile(TarLongFileMode.posix );

        File testTarArchive = getTestFile( "target/output/archived-file-set/input.tar" );
        for (int i = 0; i < 100; i++) {
            tarArchiver.addFile( getTestFile( "src/test/resources/manifests/manifest1.mf" ), "file" + i );
        }
        tarArchiver.setDestFile( testTarArchive );
        tarArchiver.createArchive();

        ZipArchiver zipArchiver = (ZipArchiver) lookup( Archiver.ROLE, "zip");
        zipArchiver.setDestFile( getTestFile("target/output/archived-file-set/archive.zip") );
        zipArchiver.addArchivedFileSet( new DefaultArchivedFileSet( testTarArchive ) );
        zipArchiver.createArchive();
    }
}
```

The exception is:
```
org.codehaus.plexus.archiver.ArchiverException: Problem creating zip: Execution exception (and the archive is probably corrupt but I could not delete it)
	at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:982)
	at org.codehaus.plexus.archiver.AddArchivedFileSetTest.testTarToZip(AddArchivedFileSetTest.java:38)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
Caused by: java.io.IOException: Execution exception
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver.close(AbstractZipArchiver.java:799)
	at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:969)
	... 20 more
Caused by: java.lang.RuntimeException: java.io.IOException: Error detected parsing the header
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver$1.get(AbstractZipArchiver.java:491)
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver$2.get(AbstractZipArchiver.java:599)
	at org.apache.commons.compress.archivers.zip.ZipArchiveEntryRequest.getPayloadStream(ZipArchiveEntryRequest.java:62)
	at org.apache.commons.compress.archivers.zip.ScatterZipOutputStream.addArchiveEntry(ScatterZipOutputStream.java:95)
	at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator$2.call(ParallelScatterZipCreator.java:175)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Error detected parsing the header
	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.getNextTarEntry(TarArchiveInputStream.java:280)
	at org.codehaus.plexus.archiver.tar.TarFile.findEntry(TarFile.java:187)
	at org.codehaus.plexus.archiver.tar.TarFile.getInputStream(TarFile.java:162)
	at org.codehaus.plexus.archiver.tar.TarFile.getInputStream(TarFile.java:134)
	at org.codehaus.plexus.archiver.tar.TarResource.getContents(TarResource.java:64)
	at org.codehaus.plexus.components.io.resources.Deferred.getContents(Deferred.java:60)
	at org.codehaus.plexus.components.io.resources.proxy.ResourceInvocationHandler.invoke(ResourceInvocationHandler.java:62)
	at com.sun.proxy.$Proxy3.getContents(Unknown Source)
	at org.codehaus.plexus.archiver.ArchiveEntry.getInputStream(ArchiveEntry.java:126)
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver$1.get(AbstractZipArchiver.java:487)
	... 8 more
Caused by: java.lang.IllegalArgumentException: Invalid byte 32 at offset 6 in '100644 00000' len=12
	at org.apache.commons.compress.archivers.tar.TarUtils.parseOctal(TarUtils.java:137)
	at org.apache.commons.compress.archivers.tar.TarUtils.parseOctalOrBinary(TarUtils.java:168)
	at org.apache.commons.compress.archivers.tar.TarArchiveEntry.parseTarHeader(TarArchiveEntry.java:1013)
	at org.apache.commons.compress.archivers.tar.TarArchiveEntry.parseTarHeader(TarArchiveEntry.java:996)
	at org.apache.commons.compress.archivers.tar.TarArchiveEntry.<init>(TarArchiveEntry.java:326)
	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.getNextTarEntry(TarArchiveInputStream.java:278)
	... 17 more
```

It's interesting that every time the result of the execution (the number of files processed before the exception is thrown) is different. That makes me wonder if this is related to some concurrent access to the tar file. I tried to hard code the number of threads to 1 (`zOut = new ConcurrentJarCreator(1)`) but it does not help.

--- a/src/main/java/org/codehaus/plexus/archiver/ArchiveEntry.java
+++ b/src/main/java/org/codehaus/plexus/archiver/ArchiveEntry.java
@@ -60,6 +60,8 @@ public class ArchiveEntry
 
     private PlexusIoResourceAttributes attributes;
 
+    private final boolean addSynchronously;
+
     /**
      * @param name     the filename as it will appear in the archive. This is platform-specific
      *                 normalized with File.separatorChar
@@ -92,6 +94,8 @@ public class ArchiveEntry
 
         this.mode = permissions == -1 ? permissions : ( permissions & UnixStat.PERM_MASK ) |
             ( type == FILE ? UnixStat.FILE_FLAG : type == SYMLINK ? UnixStat.LINK_FLAG : UnixStat.DIR_FLAG );
+
+        this.addSynchronously = ( collection != null && !collection.isConcurrentAccessSupported() );
     }
 
     /**
@@ -152,6 +156,17 @@ public class ArchiveEntry
         return ( ( type == FILE ? Archiver.DEFAULT_FILE_MODE
             : type == SYMLINK ? Archiver.DEFAULT_SYMLILNK_MODE : Archiver.DEFAULT_DIR_MODE ) & UnixStat.PERM_MASK ) |
             ( type == FILE ? UnixStat.FILE_FLAG : type == SYMLINK ? UnixStat.LINK_FLAG : UnixStat.DIR_FLAG );
+    }
+
+    /**
+     * Indicates if this entry should be added to the archive synchronously
+     * before adding the next entry and/or accessing the next entry of {@link ResourceIterator}. 
+     * 
+     * @return {@code true} if this entry should be added synchronously
+     */
+    public boolean shouldAddSynchronously()
+    {
+        return addSynchronously;
     }
 
     public static ArchiveEntry createFileEntry( String target, PlexusIoResource resource, int permissions,

--- a/src/main/java/org/codehaus/plexus/archiver/jar/JarArchiver.java
+++ b/src/main/java/org/codehaus/plexus/archiver/jar/JarArchiver.java
@@ -333,7 +333,7 @@ public class JarArchiver
 
         ByteArrayInputStream bais = new ByteArrayInputStream( baos.toByteArray() );
         super.zipFile( createInputStreamSupplier( bais ), zOut, MANIFEST_NAME, System.currentTimeMillis(), null,
-                       DEFAULT_FILE_MODE, null );
+                       DEFAULT_FILE_MODE, null, false );
         super.initZipOutputStream( zOut );
     }
 
@@ -435,7 +435,7 @@ public class JarArchiver
         ByteArrayInputStream bais = new ByteArrayInputStream( baos.toByteArray() );
 
         super.zipFile( createInputStreamSupplier( bais ), zOut, INDEX_NAME, System.currentTimeMillis(), null,
-                       DEFAULT_FILE_MODE, null );
+                       DEFAULT_FILE_MODE, null, true );
     }
 
     /**
@@ -443,7 +443,7 @@ public class JarArchiver
      */
     protected void zipFile( InputStreamSupplier is, ConcurrentJarCreator zOut, String vPath,
                             long lastModified, File fromArchive,
-                            int mode, String symlinkDestination )
+                            int mode, String symlinkDestination, boolean addInParallel )
         throws IOException, ArchiverException
     {
         if ( MANIFEST_NAME.equalsIgnoreCase( vPath ) )
@@ -464,7 +464,7 @@ public class JarArchiver
             {
                 rootEntries.addElement( vPath );
             }
-            super.zipFile( is, zOut, vPath, lastModified, fromArchive, mode, symlinkDestination );
+            super.zipFile( is, zOut, vPath, lastModified, fromArchive, mode, symlinkDestination, addInParallel );
         }
     }
 

--- a/src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java
+++ b/src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java
@@ -421,11 +421,14 @@ public abstract class AbstractZipArchiver
      * @param lastModified       last modification time for the entry.
      * @param fromArchive        the original archive we are copying this
      * @param symlinkDestination
+     * @param addInParallel      Indicates if the entry should be add in parallel.
+     *                           If set to {@code false} it is added synchronously.
+     *                           If the entry is symbolic link this parameter is ignored.
      */
     @SuppressWarnings( { "JavaDoc" } )
     protected void zipFile( InputStreamSupplier in, ConcurrentJarCreator zOut, String vPath,
                             long lastModified,
-                            File fromArchive, int mode, String symlinkDestination )
+                            File fromArchive, int mode, String symlinkDestination, boolean addInParallel )
         throws IOException, ArchiverException
     {
         getLogger().debug( "adding entry " + vPath );
@@ -446,11 +449,11 @@ public abstract class AbstractZipArchiver
                 ZipEncoding enc = ZipEncodingHelper.getZipEncoding( getEncoding() );
                 final byte[] bytes = enc.encode( symlinkDestination ).array();
                 payload = new ByteArrayInputStream( bytes );
-                zOut.addArchiveEntry( ze, createInputStreamSupplier( payload ) );
+                zOut.addArchiveEntry( ze, createInputStreamSupplier( payload ), true );
             }
             else
             {
-                zOut.addArchiveEntry( ze, wrappedRecompressor( ze, in ) );
+                zOut.addArchiveEntry( ze, wrappedRecompressor( ze, in ), addInParallel );
             }
         }
     }
@@ -502,7 +505,8 @@ public abstract class AbstractZipArchiver
         };
         try
         {
-            zipFile( in, zOut, vPath, resource.getLastModified(), null, entry.getMode(), symlinkTarget );
+            zipFile( in, zOut, vPath, resource.getLastModified(), null, entry.getMode(), symlinkTarget,
+                !entry.shouldAddSynchronously() );
         }
         catch ( IOException e )
         {
@@ -583,7 +587,7 @@ public abstract class AbstractZipArchiver
 
             if ( !isSymlink )
             {
-                zOut.addArchiveEntry( ze, createInputStreamSupplier( new ByteArrayInputStream( "".getBytes() ) ) );
+                zOut.addArchiveEntry( ze, createInputStreamSupplier( new ByteArrayInputStream( "".getBytes() ) ), true );
             }
             else
             {
@@ -591,7 +595,7 @@ public abstract class AbstractZipArchiver
                 ZipEncoding enc = ZipEncodingHelper.getZipEncoding( encodingToUse );
                 final byte[] bytes = enc.encode( symlinkDestination ).array();
                 ze.setMethod( ZipArchiveEntry.DEFLATED );
-                zOut.addArchiveEntry( ze, createInputStreamSupplier( new ByteArrayInputStream( bytes ) ) );
+                zOut.addArchiveEntry( ze, createInputStreamSupplier( new ByteArrayInputStream( bytes ) ), true );
             }
         }
     }

--- a/src/main/java/org/codehaus/plexus/archiver/zip/ConcurrentJarCreator.java
+++ b/src/main/java/org/codehaus/plexus/archiver/zip/ConcurrentJarCreator.java
@@ -37,6 +37,7 @@ public class ConcurrentJarCreator {
     private final ScatterZipOutputStream directories;
     private final ScatterZipOutputStream metaInfDir;
     private final ScatterZipOutputStream manifest;
+    private final ScatterZipOutputStream synchronousEntries;
 
     private final ParallelScatterZipCreator parallelScatterZipCreator;
     private long zipCloseElapsed;
@@ -61,6 +62,7 @@ public class ConcurrentJarCreator {
         directories = createDeferred(defaultSupplier);
         manifest = createDeferred(defaultSupplier);
         metaInfDir = createDeferred( defaultSupplier );
+        synchronousEntries = createDeferred( defaultSupplier );
 
         parallelScatterZipCreator = new ParallelScatterZipCreator(Executors.newFixedThreadPool(nThreads), defaultSupplier);
     }
@@ -72,10 +74,12 @@ public class ConcurrentJarCreator {
      *
      * @param zipArchiveEntry The entry to add. Compression method
      * @param source          The source input stream supplier
+     * @param addInParallel   Indicates if the entry should be add in parallel.
+     *                        If set to {@code false} the entry is added synchronously.
      * @throws java.io.IOException
      */
-
-    public void addArchiveEntry(final ZipArchiveEntry zipArchiveEntry, final InputStreamSupplier source) throws IOException {
+    public void addArchiveEntry(final ZipArchiveEntry zipArchiveEntry, final InputStreamSupplier source,
+            final boolean addInParallel) throws IOException {
         final int method = zipArchiveEntry.getMethod();
         if (method == -1) throw new IllegalArgumentException("Method must be set on the supplied zipArchiveEntry");
         if (zipArchiveEntry.isDirectory() && !zipArchiveEntry.isUnixSymlink()) {
@@ -93,8 +97,10 @@ public class ConcurrentJarCreator {
             if (zipArchiveEntry.isDirectory()) zipArchiveEntry.setMethod(ZipEntry.STORED);
             manifest.addArchiveEntry(createZipArchiveEntryRequest(zipArchiveEntry, createInputStreamSupplier(payload)));
             payload.close();
-        } else {
+        } else if (addInParallel) {
             parallelScatterZipCreator.addArchiveEntry(zipArchiveEntry, source);
+        } else {
+            synchronousEntries.addArchiveEntry( createZipArchiveEntryRequest( zipArchiveEntry, source ) );
         }
     }
 
@@ -110,12 +116,14 @@ public class ConcurrentJarCreator {
         metaInfDir.writeTo( targetStream );
         manifest.writeTo(targetStream);
         directories.writeTo(targetStream);
+        synchronousEntries.writeTo( targetStream );
         parallelScatterZipCreator.writeTo( targetStream);
         long startAt = System.currentTimeMillis();
         targetStream.close();
         zipCloseElapsed = System.currentTimeMillis() - startAt;
         manifest.close();
         directories.close();
+        synchronousEntries.close();
     }
 
     /**

--- a/src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java
+++ b/src/test/java/org/codehaus/plexus/archiver/zip/ArchiveFileComparator.java
@@ -94,6 +94,22 @@ public class ArchiveFileComparator
         is1.close();
         is2.close();
     }
+
+    private static void assertEquals( ArchiveFile file1, TarArchiveEntry entry1,
+                                      ZipFile file2, ZipArchiveEntry entry2 )
+            throws IOException
+    {
+        Assert.assertEquals( entry1.isDirectory(), entry2.isDirectory() );
+
+        final InputStream is1 = file1.getInputStream( entry1 );
+        final InputStream is2 = file2.getInputStream( entry2 );
+        final byte[] bytes1 = IOUtil.toByteArray( is1 );
+        final byte[] bytes2 = IOUtil.toByteArray( is2 );
+        Assert.assertTrue( Arrays.equals( bytes1, bytes2 ) );
+        is1.close();
+        is2.close();
+    }
+
     private static void assertEquals( ZipFile file1, ZipArchiveEntry entry1,
                                       ZipFile file2, ZipArchiveEntry entry2 )
             throws Exception
@@ -131,6 +147,26 @@ public class ArchiveFileComparator
                 Assert.assertNotNull(ze2);
                 assertEquals(file1, ze1, file2, ze2);
 
+            }
+        } );
+        Assert.assertTrue( map2.isEmpty() );
+    }
+
+    
+    public static void assertEquals( final ArchiveFile file1, final ZipFile file2, final String prefix )
+        throws Exception
+    {
+        final Map<String,ZipArchiveEntry> map2 = getFileEntries( file2 );
+        forEachTarArchiveEntry( file1, new TarArchiveEntryConsumer()
+        {
+            public void accept( TarArchiveEntry ze1 )
+                throws IOException
+            {
+                final String name1 = ze1.getName();
+                final String name2 = prefix == null ? name1 : ( prefix + name1 );
+                ZipArchiveEntry ze2 = map2.remove( name2 );
+                Assert.assertNotNull( ze2 );
+                assertEquals( file1, ze1, file2, ze2 );
             }
         } );
         Assert.assertTrue( map2.isEmpty() );

--- a/src/test/java/org/codehaus/plexus/archiver/zip/ConcurrentJarCreatorTest.java
+++ b/src/test/java/org/codehaus/plexus/archiver/zip/ConcurrentJarCreatorTest.java
@@ -75,7 +75,7 @@ public class ConcurrentJarCreatorTest {
                         throw new RuntimeException(e);
                     }
                 }
-            });
+            }, true);
         }
 
     }

--- a/src/test/java/org/codehaus/plexus/archiver/zip/ZipArchiverTest.java
+++ b/src/test/java/org/codehaus/plexus/archiver/zip/ZipArchiverTest.java
@@ -34,6 +34,8 @@ import org.codehaus.plexus.archiver.ArchiverException;
 import org.codehaus.plexus.archiver.BasePlexusArchiverTest;
 import org.codehaus.plexus.archiver.UnArchiver;
 import org.codehaus.plexus.archiver.UnixStat;
+import org.codehaus.plexus.archiver.tar.TarArchiver;
+import org.codehaus.plexus.archiver.tar.TarFile;
 import org.codehaus.plexus.archiver.util.ArchiveEntryUtils;
 import org.codehaus.plexus.archiver.util.DefaultArchivedFileSet;
 import org.codehaus.plexus.archiver.util.DefaultFileSet;
@@ -670,6 +672,36 @@ public class ZipArchiverTest
 
         final ZipFile cmp1 = new ZipFile( zipFile );
         final ZipFile cmp2 = new ZipFile( zipFile2 );
+        ArchiveFileComparator.assertEquals( cmp1, cmp2, "prfx/" );
+        cmp1.close();
+        cmp2.close();
+    }
+
+    public void testZipNonConcurrentResourceCollection()
+        throws Exception
+    {
+        final File tarFile = getTestFile( "target/output/zip-non-concurrent.tar" );
+        TarArchiver tarArchiver = (TarArchiver) lookup( Archiver.ROLE, "tar" );
+        tarArchiver.setDestFile( tarFile );
+        // We're testing concurrency issue so we need large amount of files
+        for (int i = 0; i < 100; i++)
+        {
+            tarArchiver.addFile( getTestFile( "src/test/resources/manifests/manifest1.mf" ),
+                "manifest1.mf" + i );
+            // Directories are added separately so let's test them too
+            tarArchiver.addFile( getTestFile( "src/test/resources/manifests/manifest2.mf" ),
+                "subdir" + i + "/manifest2.mf" );
+        }
+        tarArchiver.createArchive();
+
+        final File zipFile = new File( "target/output/zip-non-concurrent.zip" );
+        ZipArchiver zipArchive = getZipArchiver( zipFile );
+        zipArchive.addArchivedFileSet( tarFile, "prfx/" );
+        zipArchive.setEncoding( "UTF-8" );
+        zipArchive.createArchive();
+
+        final TarFile cmp1 = new TarFile( tarFile );
+        final ZipFile cmp2 = new ZipFile( zipFile );
         ArchiveFileComparator.assertEquals( cmp1, cmp2, "prfx/" );
         cmp1.close();
         cmp2.close();

